self-core-impl/src/main/java/com/selfxdsd/core/contracts/invoices/StoredPayment.java

```diff
@@ -0,0 +1,137 @@
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.selfxdsd.core.contracts.invoices;
+
+import com.selfxdsd.api.Invoice;
+import com.selfxdsd.api.Payment;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+/**
+ * A Payment stored in Self.
+ * @author Ali FELLAHI (fellahi.ali@gmail.com)
+ * @version $Id$
+ * @since 0.0.67
+ */
+public final class StoredPayment implements Payment {
+
+    /**
+     * Invoice to which this payment made.
+     */
+    private final Invoice invoice;
+
+    /**
+     * Successful Payment transaction id.
+     */
+    private final String transactionId;
+
+    /**
+     * Payment timestamp.
+     */
+    private final LocalDateTime paymentTime;
+
+    /**
+     * Payment amount.
+     */
+    private final BigDecimal value;
+
+    /**
+     * Payment status.
+     */
+    private final String status;
+
+    /**
+     * Payment failure reason.
+     */
+    private final String failReason;
+
+    /**
+     * Ctor.
+     * @param invoice The Invoice to which this payment made.
+     * @param transactionId Successful payment transaction id.
+     * @param paymentTime Payment time.
+     * @param value Payment amount.
+     * @param status Payment status (FAILED or SUCCESSFUL).
+     * @param failReason Payment failure reason.
+     */
+    public StoredPayment(final Invoice invoice,
+                         final String transactionId,
+                         final LocalDateTime paymentTime,
+                         final BigDecimal value,
+                         final String status,
+                         final String failReason
+    ) {
+        this.invoice = invoice;
+        this.transactionId = transactionId;
+        this.paymentTime = paymentTime;
+        this.value = value;
+        this.status = status;
+        this.failReason = failReason;
+    }
+
+    @Override
+    public Invoice invoice() {
+        return this.invoice;
+    }
+
+    @Override
+    public String transactionId() {
+        return this.transactionId;
+    }
+
+    @Override
+    public LocalDateTime paymentTime() {
+        return this.paymentTime;
+    }
+
+    @Override
+    public BigDecimal value() {
+        return this.value;
+    }
+
+    @Override
+    public String status() {
+        return this.status;
+    }
+
+    @Override
+    public String failReason() {
+        return this.failReason;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(
+            this.invoice.invoiceId(),
+            this.paymentTime()
```

Comment:
@fellahi-ali paymentTime and transactionId should be enough, no need for the invoice's ID.

Reply:
@amihaiemil I didn't use transactionId because it could be null if the Payment failed?

Reply:
@fellahi-ali right, but we need to use it. Let's just use it and make sure it will be empty string, not null.

Reply:
@amihaiemil my concern here is that if we remove the `invoiceId` from hashCode() and equals() than all failed transactions (i.e: with empty `transactionId`) with same `paymentTime`  would be equal.

Reply:
@fellahi-ali right. Well, the paymentTime will be a timestamp so very kittle chance to have 2 Payments performes at the same millisecond.

But you're right, let's leave it there.